## Recommendation Engine

Movies recommendation system

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("movies_metadata.csv")
df.head()

c:\users\aksha\appdata\local\programs\python\python39\lib\site-packages\IPython\core\interactiveshell.py:3155: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [3]:
df.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

In [4]:
columns_to_keep = ['adult', 'genres', 'id', 'overview', 'title']
df = df[columns_to_keep]
df.head()

adult                                             genres     id  \
0  False  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  False  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  False  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  False  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4  False                     [{'id': 35, 'name': 'Comedy'}]  11862   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         title  
0                    Toy Story  
1                      Jumanji  
2             Grumpier Old Men  
3            Waiting to Exhale  
4  Father of the Bride Part II

I would like to create a content base recommender for which the plot given in 'overview' is important. Let's check for NaN values.

In [5]:
df.isna().sum()

adult         0
genres        0
id            0
overview    954
title         6
dtype: int64

I can drop these 954 rows but it's a considerable amount of data, I think it would be better to replace it with blank space(string).

Before that I would like to save this new dataset.

In [6]:
df.to_csv('metadata-smaller.csv', index=False)

In [7]:
df = pd.read_csv('metadata-smaller.csv')
print(df.shape)
print("\n\n")
df.head()

(45466, 5)





adult                                             genres     id  \
0  False  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  False  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  False  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  False  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4  False                     [{'id': 35, 'name': 'Comedy'}]  11862   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         title  
0                    Toy Story  
1                      Jumanji  
2             Grumpier Old Men  
3            Waiting to Exhale  
4  Father of the Bride Part II

I will now use TF-IDF on the overview column to get the word vector (which will also make the dataset smaller in memory hopefully)

- Import TfIdfVectorizer from scikit-learn
- Define a TF-IDF Vectorizer Object. Remove all english stop words such as 'the', 'a'

In [8]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(stop_words='english')

We already discussed that I will replace all the NaNs with an empty string

In [9]:
df['overview'] = df['overview'].fillna('')
df.isna().sum()

adult       0
genres      0
id          0
overview    0
title       6
dtype: int64

Far Better, now we can vectorize it without any issues

In [10]:
tfidf_matrix = tfidf.fit_transform(df['overview'])
tfidf_matrix.shape

(45466, 75827)

The above shape tells us that 45466 movies gave us 75827 different words

I will be using the <code>cosine-similarity</code> to calculate a numeric quantity that denotes the similarity between two movies. i use the cosine similarity score since it is independent of magnitude and is relatively easy and fast to calculate (especially when used in conjunction with TF-IDF scores, which will be explained later)

We have used the TF-IDF vectorizer, calculating the dot product between each vector will directly give us the cosine similarity score. <br>We will use sklearn's <code>linear_kernel()</code> instead of <code>cosine_similarities()</code> since it is faster.

## Computationally intensive task

In [13]:
from sklearn.metrics.pairwise import linear_kernel
# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [14]:
cosine_sim.shape

(45466, 45466)

This returns a matrix of shape <code>45466x45466</code>, which means each movie <code>overview</code> cosine similarity score with every other movie <code>overview</code>. Hence, each movie will be a <code>1x45466</code> column vector where each column will be a similarity score with each movie.

Now, let's make the title of the movies as index while preserving the original index <br>Why do we preserve the original index?<br>Well we will need a way to tally the title of the movie with it's corresponding overview.

In [16]:
indices = pd.Series(df.index, index=df['title']).drop_duplicates()
indices[:10]

title
Toy Story                      0
Jumanji                        1
Grumpier Old Men               2
Waiting to Exhale              3
Father of the Bride Part II    4
Heat                           5
Sabrina                        6
Tom and Huck                   7
Sudden Death                   8
GoldenEye                      9
dtype: int64

- Get the index of the movie given its title.
- Get the list of cosine similarity scores for that particular movie with all movies. Convert it into a list of tuples where the first element is its position, and the second is the similarity score.
- Sort the aforementioned list of tuples based on the similarity scores; that is, the second element.
- Get the top 10 elements of this list. Ignore the first element as it refers to self (the movie most similar to a particular movie is the movie itself).
- Return the titles corresponding to the indices of the top elements.

In [19]:
# Function that takes in movie title as input and outputs most similar movies
def get_recommendations(title, cosine_sim=cosine_sim):
    # Get the index of the movie that matches the title
    idx = indices[title]

    # Get the pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return df['title'].iloc[movie_indices]

Test it!!

In [20]:
get_recommendations('The Dark Knight Rises')

12481                                      The Dark Knight
150                                         Batman Forever
1328                                        Batman Returns
15511                           Batman: Under the Red Hood
585                                                 Batman
21194    Batman Unmasked: The Psychology of the Dark Kn...
9230                    Batman Beyond: Return of the Joker
18035                                     Batman: Year One
19792              Batman: The Dark Knight Returns, Part 1
3095                          Batman: Mask of the Phantasm
Name: title, dtype: object

In [21]:
get_recommendations('The Godfather')

1178               The Godfather: Part II
44030    The Godfather Trilogy: 1972-1990
1914              The Godfather: Part III
23126                          Blood Ties
11297                    Household Saints
34717                   Start Liquidation
10821                            Election
38030            A Mother Should Be Loved
17729                   Short Sharp Shock
26293                  Beck 28 - Familjen
Name: title, dtype: object

So we can see what's happening here, these are recommendations but not the type we want. We would like a more sophisticated approach.

In [22]:
# Load keywords and credits
credits = pd.read_csv('credits.csv')
keywords = pd.read_csv('keywords.csv')

In [23]:
credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [24]:
keywords.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [26]:
# Remove rows with bad IDs.
# This is derived from dataset details
df = df.drop([19730, 29503, 35587])

We have three different dataframes at this point, we need to merge these.

In [28]:
# Convert IDs to int. Required for merging
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
df['id'] = df['id'].astype('int')

In [29]:
# Merge keywords and credits into main dataframe
df = df.merge(credits, on='id')
df = df.merge(keywords, on='id')

In [44]:
df = df.drop('overview', axis=1)

In [30]:
df.head()

adult                                             genres     id  \
0  False  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  False  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  False  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  False  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4  False                     [{'id': 35, 'name': 'Comedy'}]  11862   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                      overview_TFIDF  \
0    (0, 17764)\t0.13483149538639247\n  (0, 4388)...   
1    (0, 17764)\t0.13483149538639247\n  (0, 4388)...   
2    (0, 17764)\t0.13483149538639247\n  (0, 4388)...   
3    (0, 17764)\t0.13483149538639247\n  (0, 4388)...   
4    (0, 17764)\t0.13483149538639247\n  (0, 4388)...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...  
3  [{'id': 818, 'name': 'based on novel'}, {'id':...  
4  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

Above dataset has many stringified list, we need to convert it into python objects

In [32]:
# Parse the stringified features into their corresponding python objects
from ast import literal_eval

features = ['cast', 'crew', 'keywords', 'genres']
for feature in features:
    df[feature] = df[feature].apply(literal_eval)

In [34]:
df.head()

adult                                             genres     id  \
0  False  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  False  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  False  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  False  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4  False                     [{'id': 35, 'name': 'Comedy'}]  11862   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': 'Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...  
3  [{'id': 818, 'name': 'based on novel'}, {'id':...  
4  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

Get the director's name from the crew feature. If the director is not listed, return NaN

In [35]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

We will write a function that will return the top 4 elements or the entire list, whichever is more. Here the list refers to the cast, keywords, and genres.

In [37]:
def get_list(x):
    if isinstance(x, list):
        names = [i['name'] for i in x]
        #Check if more than 4 elements exist. If yes, return only first four. If no, return entire list.
        if len(names) > 4:
            names = names[:4]
        return names

    #Return empty list in case of missing/malformed data
    return []

In [38]:
# Define new director, cast, genres and keywords features that are in a suitable form.
df['director'] = df['crew'].apply(get_director)

features = ['cast', 'keywords', 'genres']
for feature in features:
    df[feature] = df[feature].apply(get_list)

In [39]:
# Print the new features of the first 3 films
df[['title', 'cast', 'director', 'keywords', 'genres']].head(3)

title                                               cast  \
0         Toy Story    [Tom Hanks, Tim Allen, Don Rickles, Jim Varney]   
1           Jumanji  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...   
2  Grumpier Old Men  [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...   

        director                                           keywords  \
0  John Lasseter                   [jealousy, toy, boy, friendship]   
1   Joe Johnston  [board game, disappearance, based on children'...   
2  Howard Deutch  [fishing, best friend, duringcreditsstinger, o...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]

In [40]:
df.head()

adult                        genres     id  \
0  False   [Animation, Comedy, Family]    862   
1  False  [Adventure, Fantasy, Family]   8844   
2  False             [Romance, Comedy]  15602   
3  False      [Comedy, Drama, Romance]  31357   
4  False                      [Comedy]  11862   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                                cast  \
0    [Tom Hanks, Tim Allen, Don Rickles, Jim Varney]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...   
2  [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...   
3  [Whitney Houston, Angela Bassett, Loretta Devi...   
4  [Steve Martin, Diane Keaton, Martin Short, Kim...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...   

                                            keywords         director  
0                   [jealousy, toy, boy, friendship]    John Lasseter  
1  [board game, disappearance, based on children'...     Joe Johnston  
2  [fishing, best friend, duringcreditsstinger, o...    Howard Deutch  
3  [based on novel, interracial relationship, sin...  Forest Whitaker  
4          [baby, midlife crisis, confidence, aging]    Charles Shyer

The next step would be to convert the names and keyword instances into lowercase and strip all the spaces between them.<br>

Removing the spaces between words is an important preprocessing step. It is done so that our vectorizer doesn't count the Johnny of "Johnny Depp" and "Johnny Galecki" as the same. After this processing step, the aforementioned actors will be represented as "johnnydepp" and "johnnygalecki" and will be distinct to our vectorizer.

In [41]:
# Function to convert all strings to lower case and strip names of spaces
def clean_data(x):
    if isinstance(x, list):
        return [str.lower(i.replace(" ", "")) for i in x]
    else:
        #Check if director exists. If not, return empty string
        if isinstance(x, str):
            return str.lower(x.replace(" ", ""))
        else:
            return ''

In [42]:
# Apply clean_data function to your features.
features = ['cast', 'keywords', 'director', 'genres']

for feature in features:
    df[feature] = df[feature].apply(clean_data)

In [43]:
df[features].head(3)

cast  \
0        [tomhanks, timallen, donrickles, jimvarney]   
1  [robinwilliams, jonathanhyde, kirstendunst, br...   
2  [waltermatthau, jacklemmon, ann-margret, sophi...   

                                            keywords      director  \
0                   [jealousy, toy, boy, friendship]  johnlasseter   
1  [boardgame, disappearance, basedonchildren'sbo...   joejohnston   
2  [fishing, bestfriend, duringcreditsstinger, ol...  howarddeutch   

                         genres  
0   [animation, comedy, family]  
1  [adventure, fantasy, family]  
2             [romance, comedy]

We are now in a position to create our "metadata soup", which is a string that contains all the metadata that we want to feed to our vectorizer (namely actors, director and keywords).

The create_soup function will simply join all the required columns by a space. This is the final preprocessing step, and the output of this function will be fed into the word vector model.

In [54]:
def create_soup(x):
    return ' '.join(x['keywords']) + ' ' + ' '.join(x['cast']) + ' ' + x['director'] + ' ' + x['adult'] + ' ' + ' '.join(x['genres'])

In [55]:
# Create a new soup feature
df['soup'] = df.apply(create_soup, axis=1)

In [56]:
df[['soup']].head(2)

soup
0  jealousy toy boy friendship tomhanks timallen ...
1  boardgame disappearance basedonchildren'sbook ...

In [57]:
# Import CountVectorizer and create the count matrix
from sklearn.feature_extraction.text import CountVectorizer

count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(df['soup'])

In [58]:
count_matrix.shape

(46628, 88929)

In [62]:
mat = count_matrix[:23000]
mat.shape

(23000, 88929)

In [63]:
# Compute the Cosine Similarity matrix based on the count_matrix
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim2 = cosine_similarity(mat, mat)

In [ ]:
# Function that takes in movie title as input and outputs most similar movies
def get_recommendations(title, cosine_sim=cosine_sim):
    # Get the index of the movie that matches the title
    idx = indices[title]

    # Get the pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return df['title'].iloc[movie_indices]